prov/verbs/src/windows/verbs_nd_rdma.c

```diff
@@ -168,15 +171,64 @@ int rdma_destroy_id(struct rdma_cm_id *id)
 	}
 
 	id_nd = container_of(id, struct nd_cm_id, id);
+	ofi_mutex_lock(&id_nd->connect_event.base.lock);
+	ofi_mutex_lock(&id_nd->peer_event.base.lock);
+	ofi_mutex_lock(&id_nd->listen_event.base.lock);
+	if (id_nd->connect_event.base.cb_pending ||
+	    id_nd->peer_event.base.cb_pending)
+		id_nd->connector->lpVtbl->CancelOverlappedRequests(
+			id_nd->connector);
+
+	if (id_nd->listen_event.base.cb_pending)
+		id_nd->listener->lpVtbl->CancelOverlappedRequests(
+			id_nd->listener);
+
 	while (compare = ofi_atomic_get32(&id_nd->cm_events_pending)) {
 		WaitOnAddress(&id_nd->cm_events_pending, &compare,
 			      sizeof(id_nd->cm_events_pending), INFINITE);
 	}
 
-	if (id_nd->ov_file != INVALID_HANDLE_VALUE) {
-		ret = CloseHandle(id_nd->ov_file);
-		assert(ret);
+	while (id_nd->connect_event.base.cb_pending ||
+		id_nd->peer_event.base.cb_pending ||
+		id_nd->listen_event.base.cb_pending) {
+		ofi_mutex_unlock(&id_nd->connect_event.base.lock);
+		ofi_mutex_unlock(&id_nd->peer_event.base.lock);
+		ofi_mutex_unlock(&id_nd->listen_event.base.lock);
+		pthread_yield();
+		ofi_mutex_lock(&id_nd->connect_event.base.lock);
+		ofi_mutex_lock(&id_nd->peer_event.base.lock);
+		ofi_mutex_lock(&id_nd->listen_event.base.lock);
+	}
+
+	ofi_mutex_unlock(&id_nd->connect_event.base.lock);
+	ofi_mutex_unlock(&id_nd->peer_event.base.lock);
+	ofi_mutex_unlock(&id_nd->listen_event.base.lock);
+
+	if (id->channel) {
+		struct nd_event_channel *ch_nd = container_of(
+			id->channel, struct nd_event_channel, channel);
+		if (!dlistfd_empty(&ch_nd->q)) {
+			struct nd_cm_event *entry_nd;
+			struct dlist_entry *tmp;
+			dlist_foreach_container_safe(&ch_nd->q.list,
+						     struct nd_cm_event,
+						     entry_nd, entry, tmp)
+			{
+				if (entry_nd->event.id == id) {
+					dlistfd_remove(&entry_nd->entry,
+						       &ch_nd->q);
+					if (entry_nd->event.param.conn
+						    .private_data) {
+						free((void *)entry_nd->event
+							     .param.conn
+							     .private_data);
+					}
+					free(&entry_nd->event);
+				}
+			}
+		}
```

Comment:
I would move the contents of this if statement into its own function as some sort of event queue cleanup routine.  That would avoid such deep indentation.